File: sentimental-credit/credit.py
# TODO

import re
from cs50 import get_string
number = -1

# ask user for card number
while int(number) < 0:
    number = get_string('Number: ')

# luhns algorithm

# iterate over every second value in the card number, starting from the numbers second to last digit

# set sums to start at zero
sum1 = 0
sum2 = 0


for i in range(len(number) - 2, -1, -2):
    # set dig to string so we can iterate over the digit in each sum
    dig = str(2 * int(number[i]))
    sum_dig = 0
    for j in range(len(dig)):
        sum_dig = sum_dig + int(dig[j])
    sum1 = sum1 + sum_dig

# compute second sum
for i in range(len(number) - 1, -1, -2):
    sum2 = sum2 + int(number[i])

sum_tot = sum1 + sum2

if sum_tot % 10 != 0:
    print('INVALID')

else:
    # use syntax to get the first 2 numbers from 'number'
    if len(number) == 15 and (number[0:2] == '34' or number[0:2] == '37'):
        print('AMEX')

    elif len(number) == 16 and (number[0:2] == '51' or number[0:2] == '51' or number[0:2] == '53' or number[0:2] == '54' or number[0:2] == '55'):
        print('MASTERCARD')

    elif len(number) > 12 and len(number) < 17 and number[0] == '4':
        print('VISA')

    else:
        print('INVALID')

File: credit/credit.c
#include <cs50.h>
#include <stdio.h>

int main(void)
{
    //ask for card number
    long card;
    do
    {
        card = get_long("Card Number: ");
    }
    while (card < 0);

    //count length
    int i = 0;
    long cl = card;

    while (cl > 0)
    {
        cl = cl / 10;
        i++;
    }

    //now i represents the legnth of the card number

    //check if it is a valid length (has to be 13, 15 or 16)

    if (i != 13 && i != 15 && i != 16)
    {
        printf("INVALID\n");
        return (0);
    }


    //checksum

    //sum1 is the sum of all the alternating numbers starting with the last digit
    //sum2 is the sum of all the digits of the other numbers multiplied by 2
    //d1 represets the alternating numbers starting from the last digit
    //d2 represents the alternating numbers starting from the second to last digit
    //d21 and d22 are used to seperate the two digits when multiplying d2 by 2

    long cs = card;
    int sum1 = 0;
    int sum2 = 0;
    int d1 = 0;
    int d2 = 0;
    int d21 = 0;
    int d22 = 0;
    int total = 0;

    while (cs > 0)
    {
        //find last digit
        d1 = cs % 10;

        //add to sum1
        sum1 = sum1 + d1;

        //remove last digit
        cs = cs / 10;


        //find second to last digit, now the last digit
        d2 = cs % 10;

        //multiply by 2 and add the resulting numbers digits to sum 2
        d2 = d2 * 2;
        d21 = d2 / 10;
        d22 = d2 % 10;
        sum2 = sum2 + d21 + d22;

        //remove the last digit
        cs = cs / 10;

        //now back to the start of the loop to repeat for the rest of the digits in the card number


    }

    //find total to check for luhns algorithm
    total = sum1 + sum2;

    //check last digit of total is 0
    if (total % 10 != 0)
    {
        printf("INVALID\n");
        return (0);
    }

    //now check for AMEX, mastercard or VISA

    //find the first numbers from the card number

    long first2 = card;

    while (first2 > 100)
    {
        first2 = first2 / 10;
    }

    //check card type and print if valid number

    if (first2 % 10 < 6 && first2 / 10 == 5 && first2 > 0)
    {
        printf("MASTERCARD\n");
    }

    else if (first2 / 10 == 3 && (first2 % 10 == 4 || first2 % 10 == 7))
    {
        printf("AMEX\n");
    }

    else if (first2 / 10 == 4)
    {
        printf("VISA\n");
    }

    else
    {
        printf("INVALID\n");
    }


}